src/validators/decimal.rs

```diff
@@ -0,0 +1,255 @@
+                            return Err(ValError::new(ErrorType::DecimalWholeDigits { max_whole_digits }, input));
+                        }
+                    }
+                }
+            }
+        }
+
+        if let Some(multiple_of) = &self.multiple_of {
+            // fraction = (decimal / multiple_of) % 1
+            let fraction: &PyAny = unsafe {
+                let division = PyObject::from_owned_ptr_or_err(
+                    py,
+                    pyo3::ffi::PyNumber_TrueDivide(decimal.as_ptr(), multiple_of.as_ptr()),
+                )?;
+                let one = 1.to_object(py);
+                py.from_owned_ptr_or_err(pyo3::ffi::PyNumber_Remainder(division.as_ptr(), one.as_ptr()))?
+            };
+            let zero = 0.to_object(py);
+            if !fraction.eq(&zero)? {
+                return Err(ValError::new(
+                    ErrorType::MultipleOf {
+                        multiple_of: multiple_of.to_string().into(),
+                    },
+                    input,
+                ));
+            }
+        }
+
+        if let Some(le) = &self.le {
+            if !decimal.le(le)? {
+                return Err(ValError::new(
+                    ErrorType::LessThanEqual {
+                        le: Number::String(le.to_string()),
+                    },
+                    input,
+                ));
+            }
+        }
+        if let Some(lt) = &self.lt {
+            if !decimal.lt(lt)? {
+                return Err(ValError::new(
+                    ErrorType::LessThan {
+                        lt: Number::String(lt.to_string()),
+                    },
+                    input,
+                ));
+            }
+        }
+        if let Some(ge) = &self.ge {
+            if !decimal.ge(ge)? {
+                return Err(ValError::new(
+                    ErrorType::GreaterThanEqual {
+                        ge: Number::String(ge.to_string()),
+                    },
+                    input,
+                ));
+            }
+        }
+        if let Some(gt) = &self.gt {
+            if !decimal.gt(gt)? {
+                return Err(ValError::new(
+                    ErrorType::GreaterThan {
+                        gt: Number::String(gt.to_string()),
+                    },
+                    input,
+                ));
+            }
+        }
+
+        Ok(decimal.into())
+    }
+
+    fn different_strict_behavior(
+        &self,
+        _definitions: Option<&DefinitionsBuilder<CombinedValidator>>,
+        _ultra_strict: bool,
+    ) -> bool {
+        true
+    }
+
+    fn get_name(&self) -> &str {
+        Self::EXPECTED_TYPE
+    }
+
+    fn complete(&mut self, _definitions: &DefinitionsBuilder<CombinedValidator>) -> PyResult<()> {
+        Ok(())
+    }
+}
+
+pub(crate) fn create_decimal<'a>(
+    arg: &'a PyAny,
+    input: &'a impl Input<'a>,
+    decimal_type: &'a PyType,
+) -> ValResult<'a, &'a PyAny> {
+    decimal_type.call1((arg,)).map_err(|e| {
```

Comment:
Concerning this part maybe we can do the calculation in rust and set the attributes of the python class `Decimal`: `__slots__ = ('_exp','_int','_sign', '_is_special')` WDYT @davidhewitt ?

Reply:
In this PR I'm trying to keep the same logic as `pydantic` just move the implementation. I would definitely be interested in this as a future improvement. I believe @samuelcolvin had some previous thoughts on how the parsing might be done (by looking at the digits as a string), not sure if there was any implementation put somewhere?